[RTM] Replaced the BE_PAGE_OFFSET cookie with sessionStorage

Toflar opened the pull request.

Toflar committed:
Replaced the BE_PAGE_OFFSET cookie with localstorage

Toflar commented on core-bundle/src/Resources/public/core.js:
```diff
@@ -651,6 +651,7 @@ var AjaxRequest =
 	toggleFieldset: function(el, id, table) {
 		el.blur();
 		var fs = $('pal_' + id);
+		Backend.getScrollOffset();
```
During my tests I've noticed that opening and closing legends does not store the offset. But it's actually quite nice to get back to the place you've last opened or closed a legend so I've added that here :)

Toflar commented on core-bundle/src/Resources/public/core.js:
```diff
@@ -992,10 +993,50 @@ var Backend =
+		// Remove local storage item again
+		window.localStorage.removeItem('contao_backend_offset');
+
+		var chunks = item.split(/^(\d*)/);
```
I'm storing the offset plus the current window location, so the item contains something like `768https://......./contao?....`. Why do I store the current location? Well, it's to replace the difference between the `save` and the `saveNcreate`, `saveNclose` etc. buttons. Just a lot smarter 😄 The old code called `System::setCookie('BE_PAGE_OFFSET', 0, 0);` for all the buttons except for the `save` button which is the one where you stay on the same page. This makes sense, if you save (e.g. using keyboard shortcuts) and stay on the same page, you want to jump back to the location where you've last edited but you don't want to jump there if you are on a different page (which you are when you `saveNclose` etc.). So I thought about adding some data attribute to the buttons but then I noticed that it only ever makes sense not to reset, when you are on the same page. So I'm storing the `window.location` along with the offset and if they do not match, I'm not jumping there which is bascially exactly the same functionality as we have now but it works without having to configure anything at all.

Toflar commented on core-bundle/src/Resources/public/core.js:
```diff
@@ -992,10 +993,50 @@ var Backend =
+		}
+
+		// Collect additional offset
+		$$('[data-additional-scroll-offset]').each(function(el) {
```
While I've tried doing things a bit smarter I've also come up with this solution rather than the hardcoded additional `40` pixels when `<div id="pastehint">` is present on the page. This solution can work for everything, not just the paste hint and it can also be used multiple times so you can have additional scroll offset configured by 4 elements if you like.

leofeyer commented:
Great work with some clever solutions! 🦊 

I wonder if we should use session storage instead of local storage here though, because a) the scroll offset is short-term information and b) the scroll offset should not be shared between tabs. WDYT?

Toflar commented:
Yeah I was thinking about that too but I didn't know 😄 Until now, as it was a cookie, the information was shared but I agree it makes no sense because you would probably scroll to somewhere in nowhere when you work with two tabs. I'll switch to `sessionStorage` 👍

Toflar committed:
Switched to sessionStorage

Toflar committed:
Do not hardcode additional scroll pixels but be smart about it too

Toflar commented:
Done, it's `sessionStorage` now. And I've also reworked the additional scroll pixels again to automatically calculate them instead of a hardcoded height. Currently it scrolls way to much down on mobile where the `pastehint` is not displayed (but still present, so it currently adds `40` pixels).

So apart from removing the cookie the solution is now a lot better because

* it calculates the height of the additional elements automatically and only when needed
* it does not share the scroll information across tabs

😎

leofeyer commented on core-bundle/src/Resources/public/core.js:
```diff
@@ -992,10 +993,50 @@ var Backend =
+		// Remove session storage item again
+		window.sessionStorage.removeItem('contao_backend_offset');
+
+		var chunks = item.split(/^(\d*)/);
```
I think it would be more efficient to use a trailing character such as `|` so we do not need the regex, wouldn't it?

Toflar replied:
I think the regex ist pretty fast because it's super simple. Otherwise we'd need to find a character that is not allowed in an URL.

leofeyer replied:
I ran a quick test with the null byte as separator (probably not the best choice) and it was 18% faster: http://jsben.ch/hyQhS

<img width="680" alt src="https://user-images.githubusercontent.com/1192057/57124613-b4549c80-6d86-11e9-802f-80ee355796f4.png">

Toflar replied:
Let's use this then, do you want to push yourself as it was your idea anyway? :)

Toflar replied:
Not needed anymore, so I mark this as resolved.

leofeyer commented on core-bundle/src/Resources/contao/drivers/DC_Folder.php:
```diff
@@ -455,7 +455,7 @@ public function showAll()
 <input type="hidden" name="FORM_SUBMIT" value="tl_select">
 <input type="hidden" name="REQUEST_TOKEN" value="'.REQUEST_TOKEN.'">' : '').($blnClipboard ? '
-<div id="paste_hint">
+<div id="paste_hint" data-additional-scroll-offset-relevant>
```
Do you think we can short this a little bit? 🙈

Toflar replied:
I'm not a fan of nonsense abbreviations like `asor`. Nobody will know what it does. Proposals?

leofeyer replied:
Maybe `data-add-to-scroll-offset`? That would still be clear and 13 characters shorter.

Toflar replied:
Done.

Toflar committed:
Renamed data attribute

leofeyer committed:
Fix the coding style

leofeyer committed:
Remove another occurence of BE_PAGE_OFFSET

leofeyer committed:
Correctly calculate the scroll offset

leofeyer reviewed (changes requested):
There are still two bugs. The script works perfectly on the "edit" page but unfortunately not on the listing page. Watch this video before reading on:

https://www.dropbox.com/s/7tjw5orx0c886wv/scroll-offset.mov?dl=0

1) After I have clicked the "cut" icon, the URL changes, so I had to remove the `window.location != chunks[2]` check to make it work at all.

2) As you can see, the paste hint adds too much scroll offset. The element height is calculated as 60 pixels, however, the real offset is only 22 pixels.

To fix 2), we could change the data attribute to `data-add-scroll-offset="22"`, couldn't we?

Toflar commented:
>     1. After I have clicked the "cut" icon, the URL changes, so I had to remove the `window.location != chunks[2]` check to make it work at all.

But if we do that, it would not get reset for e.g. `saveNback`, right?

> As you can see, the paste hint adds too much scroll offset. The element height is calculated as 60 pixels, however, the real offset is only 22 pixels.
> To fix 2), we could change the data attribute to data-add-scroll-offset="22", couldn't we?

Well, I've removed the hardcoded pixel height of `40` because I wanted it to be smarter than now. Right now it scrolls to the wrong position too then because it just always scrolls `40` pixels. I used the scroll height, not the element height? Maybe there's a better option than `el.getScrollSize().y`? I've tried with `el.getDimensions()` and `el.getComputedSize()` etc. but they all returned `0` because the paste hint `div` itself has no height, the children do. I'm pretty sure there must be some JS element property that contains the height of all children maybe? @ausi ?

leofeyer commented:
> > 1. After I have clicked the "cut" icon, the URL changes, so I had to remove the `window.location != chunks[2]` check to make it work at all.
> 
> But if we do that, it would not get reset for e.g. `saveNback`, right?

I have only removed the code for the video. The URL check has to stay, but it has to be fixed for the listing page.

> > To fix 2), we could change the data attribute to data-add-scroll-offset="22", couldn't we?
> 
> Well, I've removed the hardcoded pixel height of `40` because I wanted it to be smarter than now. […] I'm pretty sure there must be some JS element property that contains the height of all children maybe? @ausi ?

But that would not fix my issue. I need the additional scroll offset to be **less** than the element height.

Toflar commented:
> But that would not fix my issue. I need the additional scroll offset to be **less** than the element height.

Hardcoding it to `22` would also be wrong though because in theory (not for the paste hint but maybe for any other element) when I shrink the screen size the element gets higher, right? So it would be wrong again. Can you tell **why** you want to add less pixels than the height of the element?

leofeyer commented:
We do not need to hardcode it. The attribute value can be optional; if it is set, we use it, and if not, we measure the element's height.

> Can you tell why you want to add less pixels than the height of the element?

The paste hint is positioned relatively and uses less space than calculated by `el.getScrollSize().y`. As I said, the table only moves down by 22 pixels, even though `el.getScrollSize().y` returns `60`.

Toflar committed:
Always scroll if present and reset on the submit buttons (except for the save)

asaage commented:
60px is correct for the height of #paste_hint's children (font-size+padding+margin+rotation)
but as the position is absolut it's height doesn't matter at all.
the shift is caused by the listing_container itself.
```
#paste_hint+.tl_listing_container {
    margin-top: 30px; */
}
#tl_buttons+.tl_listing_container {
    margin-top: 10px;
}
```

Toflar commented:
Okay I've reworked this here. We're now storing the offset only and I've added events on the buttons to delete the entry (except for `save`). This now works for me. Is the scroll offset still an issue for you? For me it calculates `44px` and I'm exactly at the correct position then.

leofeyer committed:
Always remove the item from the storage, even if the offset is 0

leofeyer commented:
> Is the scroll offset still an issue for you?

Unfortunately yes. 🙈

I have probably not explained well what I consider to be the problem, so here are some screenshots.

<img width="640" src="https://user-images.githubusercontent.com/1192057/57218087-bc197880-6ff4-11e9-9c8e-4af9816cbdfb.png">

The yellow line is where the cursor is. Now when I click the "cut" icon, I want the page to be at the exact same vertical position (current behavior):

<img width="640" src="https://user-images.githubusercontent.com/1192057/57218140-db180a80-6ff4-11e9-9fa3-0d1771c509e6.png">

With your changes, however, the page gets scrolled down too much:

<img width="640" src="https://user-images.githubusercontent.com/1192057/57218171-f5ea7f00-6ff4-11e9-8d67-4648abd2bce2.png">

IMHO, this can only be fixed with `data-add-scroll-offset="…"`, because as @asaage explained, the offset is not caused by the paste hint but by moving the main view down via CSS:

```css
#paste_hint+.tl_listing_container {
    margin-top: 30px;
}
```

asaage commented:
In Articles and Pages (preceeding a .tree_view) the ```#paste_hint``` looks a little off anyway (compared to ContentElements).
That's because of ```.tl_listing_container.tree_view {
    position: relative;
}```

![image](https://user-images.githubusercontent.com/4138336/57221863-f38e2200-7000-11e9-8ded-b0e3c8ff1f4e.png) 
![image](https://user-images.githubusercontent.com/4138336/57221992-5ed7f400-7001-11e9-8b26-fd8ca484cc0b.png)

Maybe it's easier to get rid of the absolut-positioning and the differing margins?

```css
#paste_hint+.tl_listing_container {
    /* margin-top: 30px; */
}
#paste_hint {
    /* position: relative; */
}
#paste_hint p {
    /* position: absolute; */
    /* top: 0; */
    /* right: 30px; */
    transform-origin: top right;
    text-align: right;
    margin-right: 30px;  /*?*/
    margin-bottom: -26px; /*?*/
}
/* untested*/
```

aschempp commented on core-bundle/src/Resources/public/core.js:
```diff
@@ -992,10 +994,53 @@ var Backend =
 	 */
 	getScrollOffset: function() {
-		document.cookie = "BE_PAGE_OFFSET=" + window.getScroll().y + "; path=" + (Contao.path || '/');
+		window.sessionStorage.setItem('contao_backend_offset', parseInt(window.getScroll().y, 10));
```
Just FYI, the `window` prefix is not necessary

Toflar replied:
I know.

aschempp commented on core-bundle/src/Resources/public/core.js:
```diff
@@ -992,10 +994,53 @@ var Backend =
+	initScrollOffset: function() {
+		// Kill the legacy cookie here; this way it can be sent by the server
+		// but it wont be resent by the client in the next request
+		Cookie.dispose('BE_PAGE_OFFSET');
```
Are you sure this works and the cookie is never "http-only"?

Toflar replied:
It won't work if someone sets the cookie `http-only`. I think the use case is just the one where people copied the code to their own `DC_*` classes or so. I guess if they did, they just copied `System::setCookie('BE_PAGE_OFFSET', 0, 0);`. We could also just adjust `System::setCookie()` and make it ignore the `BE_PAGE_OFFSET` cookie, wdyt @leofeyer?

Toflar replied:
Well of course it was not `http-only` so far. `Backend.getScrollOffset()` set the cookie so it was never `http-only`. It could've only been `http-conly` if you set it using `System::setCooke()` which - if you did so - would've lead to weird behaviour. So I guess this is good then.

aschempp commented on core-bundle/src/Resources/public/core.js:
```diff
@@ -992,10 +994,53 @@ var Backend =
+			var form = button.getParent('form');
+			if (!form) return;
+
+			button.addEvent('click', function(e) {
```
That looks reeeally hacky. If I understand correctly, we need to add the hidden element to keep track of which button was clicked? And simply handling the click event would submit "too fast" for the JS to apply?
Maybe we could use "mouseup" or similar instead? @ausi

ausi replied:
Wouldn’t the following code achieve the same?

```js
button.addEvent('click', function(e) {
	sessionStorage.removeItem('contao_backend_offset');
}
```

If you don’t call `preventDefault()` you also don’t habe to manually call `form.submit()` and the clicked button should be handled as usual?

Toflar replied:
I've tested if you don't call `preventDefault()` it just does what the default is: submit the form. It doesn't do anything else.

ausi replied:
I was not able to reproduce that, the value was successfully removed from the sessionStorage in my test.

Toflar replied:
I'll double check. It was the reason why I built this complicated solution in the first place.

Toflar replied:
Strange, it works for me too now. Removed complexity in 3535902d9f4aadea6832072e2ae6c8c49e752f0e

Toflar committed:
Support both, positive and negative pixel and percentage values in data-add-to-scroll-offset

Toflar commented:
@leofeyer there you go. You can now use

* `data-add-scroll-offset` - will just add the current scroll size to the offset.
* `data-add-scroll-offset="20"` - will add `20` pixels to the offset.
* `data-add-scroll-offset="-20"` - will subtract `20` pixels from the offset.
* `data-add-scroll-offset="20%"` - will add `20` percent of the current scroll size to the offset.
* `data-add-scroll-offset="-20%"` - will subtract `20` percent of the current scroll size from the offset.

So you should be able to configure whatever you need now 😄

Toflar committed:
Removed useless complexity

leofeyer committed:
Do not use parseInt() if the value is an integer already

leofeyer commented:
If you agree with my changes in 8b280fcf89d7970d7696d838d6d789eb97757196, I would rebase the branch and merge it now.

Toflar commented:
I do but you did not update the definitions in `DC_Table` so it should still scroll down too far?

leofeyer committed:
Limit the scroll offset of the paste hint

leofeyer commented:
Ah, you are right. 👍  Added in 22cae991d68b1780ea839e6a7cc8a3dc6cc090b4.

Toflar commented:
Excellent, this is RTM then 👍 
👋 to yet another pointless cookie 😎

leofeyer commented:
Very nice! Thank you @Toflar.
